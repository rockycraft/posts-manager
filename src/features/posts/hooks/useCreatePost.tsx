import type { State } from "@/types/State";
import { useCallback, useState } from "react";
import { useDispatch } from "react-redux";
import { createPost } from "../slices/postsSlice";
import { postPost } from "../services/postPost";
import type { Post } from "@/types/Post";

export function useCreatePost() {
	const [state, setState] = useState<State>("idle");
	const dispatch = useDispatch();

	const create = useCallback(
		async (post: Omit<Post, "id">) => {
			try {
				setState("loading");
				const response = await postPost(post);
				dispatch(createPost(response));
				setState("success");
			} catch (error) {
				setState("error");
			}
		},
		[dispatch],
	);

	return { create, state };
}
