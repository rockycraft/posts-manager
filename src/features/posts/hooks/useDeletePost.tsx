import type { State } from "@/types/State";
import { useState } from "react";
import { deletePost as deletePostService } from "../services/deletePost";
import { useDispatch } from "react-redux";
import { removePost } from "../slices/postsSlice";
import { Post } from "@/types/Post";

export function useDeletePost() {
	const [state, setState] = useState<State>("idle");
	const dispatch = useDispatch();

	async function deletePost(id: Post["id"]) {
		try {
			setState("loading");
			//@ts-ignore
			const response = await deletePostService(id);
			dispatch(removePost(response.id));
			setState("success");
		} catch (error) {
			console.error(error);
			setState("error");
		}
	}

	return { deletePost, state };
}
