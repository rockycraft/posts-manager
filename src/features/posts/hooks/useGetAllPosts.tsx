import type { RootState } from "@/store";
import { useCallback, useEffect, useRef, useState } from "react";
import { shallowEqual, useDispatch, useSelector } from "react-redux";
import { getPosts } from "../services/getPosts";
import { addPosts } from "../slices/postsSlice";
import type { State } from "@/types/State";

export function useGetAllPosts() {
	const [state, setState] = useState<State>("idle");
	const initialized = useRef(false);
	const posts = useSelector(
		(state: RootState) => state.posts.posts,
		shallowEqual,
	);
	const dispatch = useDispatch();

	const getAllPosts = useCallback(async () => {
		try {
			setState("loading");
			const response = await getPosts();
			dispatch(addPosts(response));
			setState("success");
		} catch (error) {
			setState("error");
			console.error(error);
		}
	}, [dispatch]);

	useEffect(() => {
		if (!initialized.current) {
			initialized.current = true;

			getAllPosts();
		}
	}, [getAllPosts]);

	return { data: posts, state, getAllPosts };
}
