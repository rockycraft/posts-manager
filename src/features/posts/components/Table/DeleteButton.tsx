import { Button } from "@chakra-ui/react";
import { useDeletePost } from "../../hooks/useDeletePost";
import { HiTrash } from "react-icons/hi";
import { useCallback } from "react";
import type { Post } from "@/types/Post";

type DeleteButtonProps = {
	id: Post["id"];
};

export function DeleteButton({ id }: DeleteButtonProps) {
	const { deletePost } = useDeletePost();

	const handleClick = useCallback(() => {
		deletePost(id);
	}, [deletePost, id]);

	return (
		<Button
			data-testid="delete-post=button"
			colorPalette="red"
			onClick={handleClick}
		>
			<HiTrash />
			Eliminar
		</Button>
	);
}
