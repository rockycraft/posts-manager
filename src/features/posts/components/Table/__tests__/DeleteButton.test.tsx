import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import { DeleteButton } from "../DeleteButton";
import { useDeletePost } from "../../../hooks/useDeletePost";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { Provider } from "@/components/ui/provider";

Object.defineProperty(window, "matchMedia", {
	writable: true,
	value: vi.fn().mockImplementation((query) => ({
		matches: false,
		media: query,
		onchange: null,
		addListener: vi.fn(), // deprecated
		removeListener: vi.fn(), // deprecated
		addEventListener: vi.fn(),
		removeEventListener: vi.fn(),
		dispatchEvent: vi.fn(),
	})),
});

vi.mock("../../../hooks/useDeletePost");

describe("DeleteButton component", () => {
	const mockDeletePost = vi.fn();

	beforeEach(() => {
		vi.clearAllMocks();
		// @ts-ignore
		useDeletePost.mockReturnValue({
			deletePost: mockDeletePost,
		});
	});

	it("should render the delete button", () => {
		render(
			<Provider>
				<DeleteButton id="1" />
			</Provider>,
		);

		expect(screen.getByTestId("delete-post=button")).toBeInTheDocument();
	});

	it("should call deletePost with the correct id when clicked", () => {
		render(
			<Provider>
				<DeleteButton id="1" />
			</Provider>,
		);

		fireEvent.click(screen.getByTestId("delete-post=button"));

		expect(mockDeletePost).toHaveBeenCalledTimes(1);
		expect(mockDeletePost).toHaveBeenCalledWith("1");
	});

	it("should match snapshot", () => {
		const { asFragment } = render(
			<Provider>
				<DeleteButton id="1" />
			</Provider>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
