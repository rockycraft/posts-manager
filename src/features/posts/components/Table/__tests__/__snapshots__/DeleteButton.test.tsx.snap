// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DeleteButton component > should match snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <button
    class="chakra-button css-1uynepj"
    data-testid="delete-post=button"
    type="button"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 20 20"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M9 2a1 1 0 00-.894.553L7.382 4H4a1 1 0 000 2v10a2 2 0 002 2h8a2 2 0 002-2V6a1 1 0 100-2h-3.382l-.724-1.447A1 1 0 0011 2H9zM7 8a1 1 0 012 0v6a1 1 0 11-2 0V8zm5-1a1 1 0 00-1 1v6a1 1 0 102 0V8a1 1 0 00-1-1z"
        fill-rule="evenodd"
      />
    </svg>
    Eliminar
  </button>
</DocumentFragment>
`;
